domain/resource/service/resource.go

```diff
@@ -173,105 +189,193 @@ func (s *Service) GetResource(
-//   - [resourceerrors.ResourceNotFound] if the specified resource doesn't exist
-//   - [resourceerrors.UnitNotFound] if the specified unit doesn't exist
-func (s *Service) SetUnitResource(
-	ctx context.Context,
-	args resource.SetUnitResourceArgs,
-) (resource.SetUnitResourceResult, error) {
-	if err := args.UnitUUID.Validate(); err != nil {
-		return resource.SetUnitResourceResult{}, fmt.Errorf("unit id: %w", err)
+	res, err := s.st.GetResource(ctx, resourceUUID)
+	if err != nil {
+		return errors.Errorf("getting resource: %w", err)
+	}
+
+	store, err := s.resourceStoreGetter.GetResourceStore(ctx, res.Type)
+	if err != nil {
+		return errors.Errorf("getting resource store for %s: %w", res.Type.String(), err)
 	}
-	if err := args.ResourceUUID.Validate(); err != nil {
-		return resource.SetUnitResourceResult{}, fmt.Errorf("resource id: %w", err)
+
+	storageUUID, err := store.Put(ctx, resourceUUID.String(), reader, res.Size, coreresourcestore.NewFingerprint(res.Fingerprint.Fingerprint))
+	if err != nil {
+		return errors.Errorf("putting resource %q in store: %w", res.Name, err)
 	}
-	if args.RetrievedBy != "" && args.RetrievedByType == resource.Unknown {
-		return resource.SetUnitResourceResult{},
-			fmt.Errorf("RetrievedByType cannot be unknown if RetrievedBy set: %w", resourceerrors.ArgumentNotValid)
+
+	err = s.st.StoreResource(
+		ctx,
+		resourceUUID,
+		storageUUID,
+		retrievedBy,
+		retrievedByType,
+		incrementCharmModifiedVersion,
+	)
+	if err != nil {
+		return errors.Errorf("recording stored resource %q: %w", res.Name, err)
 	}
-	return s.st.SetUnitResource(ctx, args)
+	return err
 }
 
-// OpenApplicationResource returns the details of and a reader for the resource.
+// OpenResource returns the details of and a reader for the resource.
 //
 // The following error types can be expected to be returned:
-//   - [coreerrors.NotValid] is returned if the resource.UUID is not valid.
-//   - [resourceerrors.ResourceNotFound] if the specified resource does
-//     not exist.
-func (s *Service) OpenApplicationResource(
+//   - [resourceerrors.ResourceNotFound] if the specified resource does not
+//     exist.
+//   - [resourceerrors.StoredResourceNotFound] if the specified resource is not
+//     in the resource store.
+func (s *Service) OpenResource(
 	ctx context.Context,
 	resourceUUID coreresource.UUID,
 ) (resource.Resource, io.ReadCloser, error) {
 	if err := resourceUUID.Validate(); err != nil {
-		return resource.Resource{}, nil, fmt.Errorf("resource id: %w", err)
+		return resource.Resource{}, nil, errors.Errorf("resource id: %w", err)
+	}
+
+	res, err := s.st.GetResource(ctx, resourceUUID)
+	if err != nil {
+		return resource.Resource{}, nil, err
+	}
+
+	store, err := s.resourceStoreGetter.GetResourceStore(ctx, res.Type)
+	if err != nil {
+		return resource.Resource{}, nil, errors.Errorf("getting resource store for %s: %w", res.Type.String(), err)
+	}
+
+	// TODO(aflynn) ideally this would be finding the resource via the resources
```

Comment:
Its for referential integrity. As it is implemented at the moment, its an artifact passed out of the resource store and passed into the domain for linkage. Its not an ideal implementation, but the object store doesn't currently allow you to lookup by UUID. This could be changed in the future, but for now I believe it was decided that its too big of a change for not enough gain right now.